### PR TITLE
doc: `@time f_improved` results

### PR DESCRIPTION
**Pull request by ypeels**
In the [Performance Tips example](https://github.com/JuliaLang/julia/blob/master/doc/manual/performance-tips.rst#measure-performance-with-objtime-and-pay-attention-to-memory-allocation), I was unable to obtain `112 bytes allocated` for `f_improved` (all I did was change `s = 0` to `s = 0.0`) except on **repeated** calls to `@time f_improved(10^6)`... Repeated calls also change the memory allocated for `@time f(1)`...

Here are my results, using 64-bit Julia 0.3.7 on Windows 8.1:

```
julia> function f_improved(n)
           s = 0.0
           # rest of function unchanged... is that really all it takes?
       end

julia> @time f(1)       # manual: (93784 bytes allocated)
elapsed time: 0.004260129 seconds (90824 bytes allocated)

julia> @time f(1)       # results change on repeated call
elapsed time: 2.45e-6 seconds (80 bytes allocated)

julia> @time f(10^6)    # manual: (32002136 bytes allocated)
elapsed time: 0.052752041 seconds (32003256 bytes allocated, 20.53% gc time)

julia> @time f_improved(10^6)   # (112 bytes allocated) in manual
elapsed time: 0.004055749 seconds (75344 bytes allocated)

julia> @time f_improved(10^6)
elapsed time: 0.000867914 seconds (112 bytes allocated)
```

You probably won't want to keep my `75344` number in the manual though, since my other results were all slightly different...

Other possibilities
- Am I missing something for `f_improved`?  If so, please consider "open-sourcing" it in the manual explicitly
- Maybe this is a platform-dependent (Windows-only) issue? But my other numbers for `@time f` are all pretty close...

Thank you!

### CODE REVIEW

**Comment by pao** [comment]
> except on repeated calls to @time f_improved(10^6)... Repeated calls also change the memory allocated for @time f(1)...

The first call also includes the time and allocations of JIT compiling the method. Subsequent calls are using the cached, generated machine code.

**Comment by ypeels** [comment]
Thanks for the clarification!  But it still seems unfair/misleading to compare the first call of `f` with a repeated call of `f_improved`...

**Comment by timholy** [comment]
Agreed, you should compare the second calls of both. Cutting and pasting from what you posted:
```jl
julia> @time f(10^6)    # manual: (32002136 bytes allocated)
elapsed time: 0.052752041 seconds (32003256 bytes allocated, 20.53% gc time)

julia> @time f_improved(10^6)
elapsed time: 0.000867914 seconds (112 bytes allocated)
```
Pretty good improvement, no?

**Comment by ypeels** [comment]
In that case, shouldn't the manual report the second call for `@time f(1)` too?

**Comment by StefanKarpinski** [comment]
This is my transcript for running this on the release-0.3 branch of Julia:

```jl
julia> function f(n)
           s = 0
           for i = 1:n
               s += i/2
           end
           s
       end
f (generic function with 1 method)

julia> @time f(1)
elapsed time: 0.004710563 seconds (93504 bytes allocated)
0.5

julia> @time f(10^6)
elapsed time: 0.04123202 seconds (32002136 bytes allocated)
2.5000025e11

julia> function f_improved(n)
           s = 0.0
           for i = 1:n
               s += i/2
           end
           s
       end
f_improved (generic function with 1 method)

julia> @time f_improved(1)
elapsed time: 0.003702172 seconds (78944 bytes allocated)
0.5

julia> @time f_improved(10^6)
elapsed time: 0.004313644 seconds (112 bytes allocated)
2.5000025e11
```

This matches what the documentation indicates pretty closely. If you're seeing something else, can you provide a transcript of exactly what you're running in a fresh Julia session?

**Comment by timholy** [comment]
@ypeels, the _value_ you pass to `f` does not matter; JIT compilation is dependent on types, not values. So the first call is with an `Int` (1), and the second call is with an `Int` (10^6). 

For `f_improved`, only the second call is shown.

Maybe I'm missing something, but I don't see the problem that's worrying you.

**Comment by ypeels** [comment]
@StefanKarpinski: 
Thanks for double-checking; I am getting similar results.  This is just a documentation issue.

@timholy:
On first reading, I thought the Julia code went like this:

    function f(n) ... end
    @time f(1) 
    @time f(10^6)
    function f_improved(n) ... end
    @time f(10^6)   # 112 bytes allocated on the first call!?

Since `f_improved` is "closed-source", this makes it seem like there is some undocumented magic in `f_improved` that makes its memory footprint so small... especially when the documentation points out memory usage explicitly:

> note that an improved version of this function allocates no memory

I kind of see now that the unshown call to `f_improved(1)` (or similar) is foreshadowed by the sentence,

>  You should not take the results of this run seriously.

But I still think it would be clearer either to make the first call to `f_improved` explicit, or maybe revisit the example explicitly in the ["Avoid changing the type of a variable"](https://github.com/JuliaLang/julia/blob/master/doc/manual/performance-tips.rst#avoid-changing-the-type-of-a-variable) section.

Sorry to all about the confusion

---
tl;dr

Instead of this pull request, how about adding the results of @StefanKarpinski 's first run to the documentation?  In parallel with `@time f`, the "teaser" could then read:

    julia> @time f_improved(1)   # first call
    elapsed time: 0.003702172 seconds (78944 bytes allocated)
    0.5

    julia> @time f_improved(10^6)
    elapsed time: 0.004313644 seconds (112 bytes allocated)
    2.5000025e11

**Comment by StefanKarpinski** [comment]
So you just want to include the first call to `f_improved`?

**Comment by timholy** [comment]
If you want to show the "warmup" call to `f_improved`, I'd have no objections. Instructions are [here](https://github.com/JuliaLang/julia/blob/master/CONTRIBUTING.md#improving-documentation).

**Comment by ypeels** [comment]
Yeah, that's it in a nutshell!  I guess the documentation claim of "thirty-fold faster" would have to be changed too... I guess I'll submit that pull request and close this one if there are no objections.

**Comment by timholy** [comment]
Just FYI, the ratio depends on the machine you run it on; when I wrote that, I was comparing second invocations and it really was thirty-fold. But if you're updating the numbers, you should also update the text correspondingly.

**Comment by ypeels** [comment]
Submitted: https://github.com/JuliaLang/julia/pull/10962

**Comment by ypeels** [comment]
Submitted new pull request and closed this one (gave wrong link earlier): https://github.com/JuliaLang/julia/pull/10967/
